src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java

```diff
@@ -269,6 +277,216 @@ private Optional<Pocket> findPocketForReload(ReloadBalanceRequest reloadRequest,
+            ? revertRequest.getUuid()
+            : UUID.randomUUID().toString();
+
+        BalanceChangeEvent changeEvent = createBalanceChangeEvent(
+            operationUuid,
+            reloadOperationToRevert.getAmountTotal(),
+            Metadata.of(revertRequest.getMetadata()),
+            now(clock),
+            balance,
+            true,
+            REVERT_RELOAD
+        );
+        changeEvent.setRevertOperationId(revertRequest.getUuidToRevert());
+
+        log.info("Revert reload mode: {}", revertRequest.getRevertReloadMode());
+        if (revertRequest.getRevertReloadMode() == null) {
+            log.error("Could not operation provide, because empty revert reload mode");
+            throw new BusinessException("error.revert.reload.mode.empty", "Empty revert reload mode");
+        } else if (revertRequest.getRevertReloadMode().equals(EXCEPTION) || revertRequest.getRevertReloadMode().equals(AS_MANY_AS_POSSIBLE_FROM_POCKETS)) {
+            revertReloadByMode(revertRequest.getRevertReloadMode(), reloadOperationToRevert.getPocketChangeEvents(), changeEvent);
+        } else if (revertRequest.getRevertReloadMode().equals(AS_MANY_AS_POSSIBLE_FROM_BALANCE)) {
+            chargingPockets(balance, reloadOperationToRevert.getAmountTotal(), changeEvent, now(clock), true, null);
+        } else {
+            log.error("Could not operation provide, because unknown revert reload mode: {}", revertRequest.getRevertReloadMode());
+            throw new BusinessException("error.revert.reload.mode.unknown", "Unknown revert reload mode");
+        }
+
+        changeEvent = balanceChangeEventRepository.save(changeEvent);
+
+        return balanceChangeEventMapper.toDto(changeEvent);
+    }
+
+    private void revertReloadByMode(RevertReloadMode revertReloadMode,
+                                    List<PocketChangeEventDto> pocketChangeEventDtoList,
+                                    BalanceChangeEvent changeEvent) {
+        List<PocketChangeEvent> pocketChangeEvents = new ArrayList<>();
+
+        for (PocketChangeEventDto pocketChangeEventDto : pocketChangeEventDtoList) {
+            Pocket pocket = pocketRepository.findOneByIdForUpdate(pocketChangeEventDto.getPocketId())
+                .orElseThrow(() -> {
+                    log.error("Not found pocket with id: {}", pocketChangeEventDto.getPocketId());
+                    return new EntityNotFoundException("Pocket with id " + pocketChangeEventDto.getPocketId() + "not found");
+                });
+
+            BigDecimal amountBefore = pocket.getAmount();
+            BigDecimal amountToProvide = pocketChangeEventDto.getAmountDelta();
+
+            boolean isChargingToNegativeRemainder = amountBefore.subtract(amountToProvide).compareTo(ZERO) < 0;
+
+            if (revertReloadMode.equals(EXCEPTION) && isChargingToNegativeRemainder) {
+                log.error("Could not revert reload, because mode: EXCEPTION and operation provide negative remainder");
```

Comment:
maybe `could not revert reload, not enough amount in pocket` ?

Reply:
Fixed